src/libsyntax/ast.rs

```diff
@@ -989,32 +1074,56 @@ impl Expr {
         }
     }
 
+    /// Attempts to deduce the type of expression.
```

Comment:
This sounds like it is doing type inference but it is really doing reparsing for diagnostics purposes.